src/ngraph/type.hpp

```diff
@@ -71,4 +80,133 @@ namespace ngraph
+    public:
+        static EnumType as_type(const std::string& name)
+        {
+            for (auto p : get().m_string_enums)
+            {
+                if (p.first == name)
+                {
+                    return p.second;
+                }
+            }
+            NGRAPH_CHECK(false, "\"", name, "\"", " is not a member of enum ", get().m_enum_name);
+        }
+
+        static std::string as_type(EnumType e)
+        {
+            for (auto p : get().m_string_enums)
+            {
+                if (p.second == e)
+                {
+                    return p.first;
+                }
+            }
+            NGRAPH_CHECK(false, " invalid member of enum ", get().m_enum_name);
+        }
+
+    private:
+        EnumNames(const std::string& enum_name,
+                  const std::vector<std::pair<std::string, EnumType>> string_enums)
+            : m_enum_name(enum_name)
+            , m_string_enums(string_enums)
+        {
+        }
+        static EnumNames<EnumType>& get();
+
+        const std::string m_enum_name;
+        std::vector<std::pair<std::string, EnumType>> m_string_enums;
+    };
+
+    template <typename Type, typename Value>
+    typename std::enable_if<std::is_convertible<Value, std::string>::value, Type>::type
+        as_type(const Value& value)
+    {
+        return EnumNames<Type>::as_type(value);
+    }
+
+    template <typename Type, typename Value>
+    typename std::enable_if<std::is_convertible<Type, std::string>::value, Type>::type
+        as_type(Value value)
+    {
+        return EnumNames<Value>::as_type(value);
+    }
+
+    class VisitorAdapter
+    {
+    public:
+        static constexpr DiscreteTypeInfo type_info{"VisitorAdapter", 0};
+        virtual ~VisitorAdapter() {}
+        virtual const DiscreteTypeInfo& get_type_info() const { return type_info; }
+        virtual std::string get_string() const = 0;
+        virtual void set_string(const std::string& value) const = 0;
+    };
+
+    template <typename Type>
+    class TypeAdapter : public VisitorAdapter
+    {
+    public:
+        operator Type&() const { return m_value; }
+    protected:
+        TypeAdapter(Type& value)
+            : m_value(value)
+        {
+        }
+        Type& m_value;
+    };
+
+    template <typename Type>
+    class EnumAdapter : public TypeAdapter<Type>
+    {
+    public:
+        EnumAdapter(Type& value)
+            : TypeAdapter<Type>(value)
+        {
+        }
+        static const DiscreteTypeInfo type_info;
+        const DiscreteTypeInfo& get_type_info() const override { return type_info; }
+        std::string get_string() const override
+        {
+            return as_type<std::string>(TypeAdapter<Type>::m_value);
+        }
+        void set_string(const std::string& value) const override
+        {
+            TypeAdapter<Type>::m_value = as_type<Type>(value);
+        }
+    };
+
+    template <typename Type>
+    class ObjectAdapter : public TypeAdapter<Type>
```

Comment:
What is the purpose of Object adapter? Will it serialise an abstract object?

Reply:
Yes. For Enums, we use the string of the enum element, for structures we need the type_info if it's variant, and the members, where the members can be done with the attribute visitor.

Reply:
Vectors are another special case for serialization; length and elements.